Source/AliveLibAE/Renderer/OpenGLRenderer.cpp

```diff
@@ -0,0 +1,1519 @@
+    short tpagey = (16 * (tPage & 0x10) + (((unsigned int)tPage >> 2) & 0x200)) + v;
+
+    // Lets prims use background texture as a src even tho we dont have vram anymore.
+    if (tpagex < 640 && tpagey < 240)
+    {
+        TextureCache * tc = GetBackgroundTextureCache();
+        tc->mUvOffset.field_0_x = tpagex;
+        tc->mIgnoreColor = true;
+        return tc;
+    }
+
+    switch (textureMode)
+    {
+    case 0: // e4Bit_0
+        tpagex += u / 4;
+        break;
+    case 1: // e8Bit_1
+        tpagex += u / 2;
+        break;
+    case 2: // e16Bit_2
+        tpagex += u;
+        break;
+    }
+
+    for (size_t i = 0; i < gRendererTextures.size(); i++)
+    {
+        TextureCache* c = &gRendererTextures[i];
+
+        if (tpagex >= c->mVramRect.x && tpagex < c->mVramRect.x + c->mVramRect.w &&
+            tpagey >= c->mVramRect.y && tpagey < c->mVramRect.y + c->mVramRect.h)
+        {
+            return c;
+        }
+    }
+
+    return 0;
+}
+
+PSX_Point Renderer_ClutToCoords(int tClut)
+{
+    int x = (tClut & 63) << 4;
+    int y = ((tClut >> 6) & 0xff);
+
+    return { (short)x,(short)y };
+}
+
+PaletteCache* Renderer_ClutToPalette(int tClut)
+{
+    short x = (tClut & 63) << 4;
+    short y = ((tClut >> 6) & 0xff);
+
+    for (size_t i = 0; i < gRendererPals.size(); i++)
+    {
+        PaletteCache* c = &gRendererPals[i];
+
+        if (x >= c->mPalPoint.field_0_x && x < (c->mPalPoint.field_0_x + c->mPalDepth) && c->mPalPoint.field_2_y == y)
+        {
+            return c;
+        }
+    }
+
+    return nullptr;
+}
+
+TextureCache* Renderer_TexFromVRam(const PSX_RECT& rect)
+{
+    for (size_t i = 0; i < gRendererTextures.size(); i++)
+    {
+        TextureCache* c = &gRendererTextures[i];
+
+        if (c->mVramRect.x == rect.x && c->mVramRect.y == rect.y)
+        {
+            return c;
+        }
+    }
+
+    return 0;
+}
+
+void Renderer_FreeTexture(PSX_Point point)
+{
+    for (size_t i = 0; i < gRendererTextures.size(); i++)
+    {
+        TextureCache* c = &gRendererTextures[i];
+
+        if (c->mVramRect.x == point.field_0_x && c->mVramRect.y == point.field_2_y)
+        {
+            glDeleteTextures(1, &c->mTextureID);
+            gRendererTextures.erase(gRendererTextures.begin() + i);
+            return;
+        }
+    }
+}
+
+void Renderer_DecodePalette(const BYTE* srcPalData, RGBAPixel* dst, int palDepth)
+{
+    const unsigned short* palShortPtr = reinterpret_cast<const unsigned short*>(srcPalData);
+    for (int i = 0; i < palDepth; i++)
+    {
+        unsigned short oldPixel = palShortPtr[i];
+
+        dst[i].R = static_cast<unsigned char>((((oldPixel >> 0) & 0x1F)) << 2);
+        dst[i].G = static_cast<unsigned char>((((oldPixel >> 5) & 0x1F)) << 2);
+        dst[i].B = static_cast<unsigned char>((((oldPixel >> 10) & 0x1F)) << 2);
+        dst[i].A = static_cast<unsigned char>((((((oldPixel) >> 15) & 0xffff)) ? 127 : 255));
+
+        if (i == 0)
+        {
+            dst[i].A = 0;
```

Comment:
wouldn't it be better to say if (pathDepth > 0)  { dst[0].A = 0; } instead of checking on every iteration?